File: Python/1036-boys_vs_girls.py
from collections import namedtuple


Student = namedtuple('Student', ['name', 'id', 'grade'])


if __name__ == '__main__':
    n_student = int(input())
    boys, girls = [], []
    for _ in range(n_student):
        name, gender, id, grade = input().split()
        grade = int(grade)
        if gender == 'F':
            girls.append(Student(name, id, grade))
        else:
            boys.append(Student(name, id, grade))
    top_girl = bottom_boy = None
    if girls:
        top_girl = max(girls, key=lambda x: x.grade)
        print(top_girl.name, top_girl.id)
    else:
        print("Absent")
    if boys:
        bottom_boy = min(boys, key=lambda x: x.grade)
        print(bottom_boy.name, bottom_boy.id)
    else:
        print("Absent")
    if top_girl and bottom_boy:
        print(top_girl.grade - bottom_boy.grade)
    else:
        print("NA")

File: C++/1036-boys_vs_girls.cc
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>


struct Student {
    std::string name, id;
    int grade;
};


bool comp(const Student &s1, const Student &s2) {
    return s1.grade < s2.grade;
}


int main() {

    int n_student;
    std::cin >> n_student;
    
    std::vector<Student> boys, girls;
    while (n_student--) {
        std::string name, id;
        char gender;
        int grade;
        std::cin >> name >> gender >> id >> grade;
        if (gender == 'M') {
            boys.push_back({name, id, grade});
        } else {
            girls.push_back({name, id, grade});
        }
    }

    Student top_girl, bottom_boy;
    if (!girls.empty()) {
        top_girl = *std::max_element(girls.begin(), girls.end(), comp);
        std::cout << top_girl.name << ' ' << top_girl.id << '\n';
    } else {
        std::cout << "Absent\n";
    }
    if (!boys.empty()) {
        bottom_boy = *std::min_element(boys.begin(), boys.end(), comp);
        std::cout << bottom_boy.name << ' ' << bottom_boy.id << '\n';
    } else {
        std::cout << "Absent\n";
    }
    if (top_girl.name != std::string() &&
        bottom_boy.name != std::string()) {
        std::cout << top_girl.grade - bottom_boy.grade << '\n';
    } else {
        std::cout << "NA\n";
    }

    return 0;
}
